### PR TITLE
testing: provide handcrank test watcher

### PR DESCRIPTION
**Pull request by tychoish**
This just pulls apart some of the util.Watcher parts to make it
possible to have a `testutil.Watch` that can produce a `WatchStore`
populated by mock events from a channel.

### CODE REVIEW

File: pkg/util/watch.go [hunk 1]
```diff
@@ -409,10 +359,72 @@ func Watch[C WatchClient[L], L metav1.ListMetaAccessor, T any, P WatchObject[T]]
...
+		if !ok {
+			return fmt.Errorf(
+				"watch %s: received update event for object %s that's not present",
+				args.Config.LogName, args.Name,
+			)
+		}
+		store.objects[args.UID] = (*T)(args.Object)
+		handlers.UpdateFunc(old, (*T)(args.Object))
+	case watch.Error:
+		panic(errors.New("unreachable code reached")) // handled above
```
Comment (sharnoff) [inline pkg/util/watch.go:420]: note `// handled above` is still handled in the original function. It's worth adding a note that `ProcessEvent` must not be called with an event with type `watch.Error`. Should also change this panic message.

File: pkg/util/watch.go [hunk 2]
```diff
@@ -409,10 +359,72 @@ func Watch[C WatchClient[L], L metav1.ListMetaAccessor, T any, P WatchObject[T]]
...
+	return store, nil
+}
+
+type ProcessEventArgs[T any, P WatchObject[T]] struct {
+	Name   string
+	UID    types.UID
+	Config WatchConfig
+	Event  watch.Event
+	Object P
 }
```
Comment (sharnoff) [inline pkg/util/watch.go:371]: I think it's probably cleaner to move the rest of the post-error-check handling into `ProcessEvent`, which allows `Name`, `UID`, and `Object` to be derived inside. With that, we probably don't need `ProcessEventArgs`?

File: pkg/util/watch.go [hunk 3]
```diff
@@ -409,10 +359,72 @@ func Watch[C WatchClient[L], L metav1.ListMetaAccessor, T any, P WatchObject[T]]
...
+	case watch.Modified:
+		old, ok := store.objects[args.UID]
+		if !ok {
+			return fmt.Errorf(
+				"watch %s: received update event for object %s that's not present",
+				args.Config.LogName, args.Name,
+			)
+		}
+		store.objects[args.UID] = (*T)(args.Object)
+		handlers.UpdateFunc(old, (*T)(args.Object))
```
Comment (sharnoff) [inline pkg/util/watch.go:418]: Just noticed we're not locking `store.mutex` here — oops. I'll open a PR to fix that. `ProcessEvent` should probably just lock `store.mutex` outside the switch statement; it'll be cleaner.

File: pkg/util/watch.go [hunk 4]
```diff
@@ -409,10 +359,72 @@ func Watch[C WatchClient[L], L metav1.ListMetaAccessor, T any, P WatchObject[T]]
...
+		if _, ok := store.objects[args.UID]; ok {
+			return fmt.Errorf(
+				"watch %s: received add event for object %s that we already have",
+				args.Config.LogName, args.Name,
+			)
+		}
+		store.objects[args.UID] = (*T)(args.Object)
+		handlers.AddFunc((*T)(args.Object), false)
+	case watch.Bookmark:
+		// Nothing to do, just serves to give us a new ResourceVersion.
```
Comment (sharnoff) [inline pkg/util/watch.go:392]: Maybe worth having a similar "don't call this function with `watch.Error`" thing for `watch.Bookmark`; this is kind of just noise here, and we can filter it out at the callsite.

**Review by sharnoff — COMMENTED** [review]
As discussed, broadly I think this is a good direction.

Some specifics: I don't like that the hand-crank-ness is available on `WatchStore` directly. Instead of `NewWatchStore`, I'd make a separate function (e.g., `NewTestWatcher`, idk) that takes `Config` and `WatchHandlerFuncs` and returns `WatchStore` and `WatchStoreTestFeeder` (name's eh, do whatever), with methods on `WatchStoreTestFeeder` that instantiate and process each type of event through the underlying `WatchStore`. Basically, our internal testing code shouldn't need to know how to construct kubernetes watch events, and our non-testing code shouldn't be able to.

It'd also be good to hook up `WatchStore.Stop()` to work with testing `WatchStore` creation. `Stop` is also currently kind of a mess (not for any reason, it's just non-idiomatic and poorly written).

Might have some more comments the second time around; I left off comments on some of the specifics that I thought might be obviated by structural changes.

File: pkg/util/watch.go (hunk 4 above)
Comment (tychoish) [reply pkg/util/watch.go:392]: I think there's merit in being exhaustive in switch statements like these. It's also the case that if we filter things out in the main loop, then this function becomes less useless as the stateless filter through which events flow.

File: pkg/util/watch.go (hunk 3 above)
Comment (tychoish) [reply pkg/util/watch.go:418]: I moved the lock to the top. I think this a decent demonstration of why #34 would be good. Amusing the handlers don't cause a deadlock, that is.

File: pkg/util/watch.go (hunk 2 above)
Comment (tychoish) [reply pkg/util/watch.go:371]: I don't understand what you mean by `post-error-checking`. 

The `goto` and loops make it hard to move more around without making a larger (and therefore riskier) cut. I think as an intermediate state.

**Comment by sharnoff** [comment]
Closing as we have no current plans to continue with this strategy for testing (and because #371 has been merged, which does finally have tests).

Can always reopen in the future.

**Pull request closed without merging**